games/dm/dm.8
.\"	$NetBSD$
.\"
.\" Copyright (c) 1987, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)dm.8	8.1 (Berkeley) 5/31/93
.\"
.Dd May 31, 1993
.Dt DM 8
.Os
.Sh NAME
.Nm dm
.Nd dungeon master
.Sh SYNOPSIS
.Nm ln
.Fl s Cm dm Ar game
.Sh DESCRIPTION
.Nm
is a program used to regulate game playing.
.Nm
expects to be invoked with the name of a game that a user wishes to play.
This is done by creating symbolic links to
.Nm ,
in the directory
.Pa /usr/games
for all of the regulated games.
The actual binaries for these games should be placed in a
.Dq hidden
directory,
.Pa /usr/games/hide ,
that may only be accessed by the
.Nm
program.
.Nm
determines if the requested game is available and, if so, runs it.
The file
.Pa /etc/dm.conf
controls the conditions under which games may be run.
.Pp
The file
.Pa /etc/nogames
may be used to
.Dq turn off
game playing.
If the file exists, no game playing is allowed; the contents of the file
will be displayed to any user requesting a game.
.Sh FILES
.Bl -tag -width /var/log/games.log -compact
.It Pa /etc/dm.conf
configuration file
.It Pa /etc/nogames
turns off game playing
.It Pa /usr/games/hide
directory of ``real'' binaries
.It Pa /var/log/games.log
game logging file
.El
.Sh SEE ALSO
.Xr dm.conf 5
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.3 tahoe .
.Sh SECURITY CONSIDERATIONS
Two issues result from
.Nm
running the games setgid
.Dq games .
First, all games that allow users to run
.Ux
commands should carefully
set both the real and effective group ids immediately before executing
those commands.
Probably more important is that
.Nm
never be setgid anything but
.Dq games
so that compromising a game will result only in
the user's ability to play games at will.
Secondly, games which previously had no reason to run setgid and which
accessed user files may have to be modified.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 6.0 to NetBSD 7.0:
	window(1): Removed from the base system in favor of tmux(1).  Use the
		pkgsrc/misc/window package from now on.  [jmmv 20120215]
	pkg_add(1): imported 20120221 version from pkgsrc [wiz 20120221]
	file(1): upgraded to 5.11 [christos 20120222]
	evbmips: Add support for Lemote Yeeloong Notebook. [nonaka 20120302]
	dbsym(8): Add option -p [bsh 20120319]
	dhcpcd(8): Import dhcpcd-5.5.5 [roy 20120323]
	ipf: upgrade to 5.1.1 with new build framework, no module and
		test support yet [christos 20120323]
	pcc: upgrade to 1.1.0.DEVEL 20120325 [plunky 20120325]
	openresolv(8): Import openresolv-3.5.1 [roy 20120327]
	dhcpcd(8): Import dhcpcd-5.5.6 [roy 20120328]
	apple: Recognize the apple disk partitioning scheme, and auto-discover
		wedges. Fix iPod disk mounting, by adding quirk.
		[christos 20120306]
	fstab: Add NAME=wedgename for fs_spec and routines to parse it. Now
		wedges can be used in /etc/fstab, even in dynamic environments
		where disks come and go and wedges get renumbered.
		[christos 20120306]
	pdisk(8): Make it work for 64 bit hosts, and make it MI.
		[christos 20120306]
	amiga: Add clockport(4) layer and a driver for A1200 on-board clockport
		a1k2cp(4). [rkujawa 20120417]
	kernel: add mpii(4), a driver for LSI Logic Fusion-MPT Message Passing
		Interface II SAS controllers.  [bouyer 20120419]
	kernel: Add support for 256 CPUs on amd64. [rmind 20120420]
	OpenSSH: Imported 6.0 [christos 20120501]
	bzip2: Imported 1.0.6 [wiz 20120507]
	amiga: Rework X-Surf driver as xsurf(4). Add support for clockport(4)
		on top of xsurf(4). [rkujawa 20120515]
	backtrace(3): Add <execinfo.h> and the backtrace family of functions
		[christos 20120526]
	kernel: Add run(4), a driver for Ralink Technology
		RT2700U/RT2800U/RT3000U USB IEEE 802.11a/b/g/n wireless
		network devices, ported from OpenBSD. [nonaka 20120530]
	bind: update to 9.9.1-P1 to fix the recent rdata exploit.
		[christos 20120604]
	empb(4): Add ELBOX Mediator PCI 1200 driver. [rkujawa 20120604]
	kernel: Add a new RLIMIT_NTHR limit to limit the number of processes
		per user and the corresponding kern.maxlwp sysctl. Expose
		kern.uidinfo.* via sysctl. [christos 20120608]
	usb(9): Added support for SMP to the USB subsytem.  By Jared D.
		McNeill and myself.  [mrg 20120609]
	kernel: {send,recv}mmsg support [christos 20120621]
	kernel: Add sysctls to avoid ipv6 DoS attacks (from OpenBSD):
		net.inet6.ip6.neighborgcthresh = 2048
		net.inet6.ip6.maxifprefixes = 16
		net.inet6.ip6.maxifdefrouters = 16
		net.inet6.ip6.maxdynroutes = 4096
		[christos 20120622]
	ata(4): added support for sata Port MultiPliers (PMP)
		[bouyer 20120702]
	kernel, libc: Add MurmurHash2 function. [rmind 20120708]
	atf(7): Import 0.16.  [jmmv 20120711]
	dhcpcd(8): Import dhcpcd-5.6.1 [roy 20120712]
	kernel: Add support for sensors to provide entropy to rnd(4)
		[pgoyette 20120715]
	tdvfb(4): Add 3Dfx Voodoo2 driver. [rkujawa 20120719]
	tdvfb(4): Add 3Dfx Voodoo Graphics (aka Voodoo1) support.
		[rkujawa 20120720]
	kernel: Change mii_statchg to take "struct ifnet *" instead of
		"device_t". [matt 20120722]
	pr(1): Add support for POSIX -f and -p options. [ginsbach 20120724]
	powerpc: Add support for FPU emulation on BookE. [matt 20120722]
	OpenSSL: Imported 1.0.1c [christos 20120726]
	kernel: safepri is dead; replaced by macro IPL_SAFEPRI. [matt 20120727]
	kernel: Add malo(4), a driver for Marvell Libertas IEEE 802.11b/g
		wireless network devices, ported from OpenBSD.
		[degroote 20120730]
	kernel, libc: Add modctl(MODCTL_EXISTS, ...) to determine if modules
		can be loaded. [jnemeth 20120803]
	modstat(8): Add -A, -a, and -e options for testing module loadability.
		[jnemeth 20120803]
	libc: Update to tzcode2012e. [christos 20120809]
	arm: Add support for BEAGLEBOARDXM (TI DM37XX cpus) boards.
		[matt 20120822]
	mfi(8): Added tagged queuing support.  [bouyer 20120823]
	mfi(8): Added support for LSI SAS2208-based controllers.
		[bouyer 20120823]
	mkdep(1): Add -P option.  [jmmv 20120826]
	share/mk: Add placeholder test programs (with some very basic test
		cases within) for common bsd.*.mk modules.  [jmmv 20120826]
	arm: Add Cortex-A9 support (a9tmr, a9wdt, armgic, armperiph).  Add
		initial MP support (can boot 2nd cpu but dies quickly).  Add
		L1 SuperSection support for ARMv6/ARMv7.  Add support for 
		BCM5301X SoC.  Add initial support for OMAP4430 and partial
		support for Freescale i.MX6.  Add new common boot/kvminit
		code.  [matt 20120831]
	dhcpcd(8): Import dhcpcd-5.6.2 [roy 20120903]
	bge(4): Add support for BCM57762, found in Apple's Thunderbolt to
		Gigabit Ethernet Adapter. PR/46961 [tsutsui 20120917]
	mfi(4): add a command pass-through ioctl, and associated COMPAT_LINUX
		support. This allows to use the LSI MegaCLI linux binary
		on a NetBSD host.  [bouyer 20120919]
	zoneinfo: Import tzdata2012f. [apb 20120927]
	kernel: add clock_nanosleep(2) [christos 20121001]
	wpa:	New wpa_supplicant(8) and hostapd(8) 1.0 [christos 20121007]

games/dm/dm.c
/*	$NetBSD$	*/

/*
 * Copyright (c) 1987, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1987, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)dm.c	8.1 (Berkeley) 5/31/93";
#else
__RCSID("$NetBSD$");
#endif
#endif /* not lint */

#include <sys/param.h>
#include <sys/file.h>
#include <sys/time.h>
#include <sys/resource.h>

#include <err.h>
#include <ctype.h>
#include <errno.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "utmpentry.h"
#include "pathnames.h"

static time_t	now;			/* current time value */
static int	priority = 0;		/* priority game runs at */
static char	*game,			/* requested game */
		*gametty;		/* from tty? */

static void c_day(const char *, const char *, const char *);
static void c_game(const char *, const char  *, const char *, const char *);
static void c_tty(const char *);
static const char *hour(int);
static double load(void);
static void nogamefile(void);
static void play(char **) __dead;
static void read_config(void);
static int users(void);

#ifdef LOG
static void logfile(void);
#endif

int
main(int argc __unused, char *argv[])
{
	char *cp;

	nogamefile();
	game = (cp = strrchr(*argv, '/')) ? ++cp : *argv;

	if (!strcmp(game, "dm"))
		exit(0);

	gametty = ttyname(0);
	unsetenv("TZ");
	(void)time(&now);
	read_config();
#ifdef LOG
	logfile();
#endif
	play(argv);
	/*NOTREACHED*/
	return (0);
}

/*
 * play --
 *	play the game
 */
static void
play(char **args)
{
	char pbuf[MAXPATHLEN];

	snprintf(pbuf, sizeof(pbuf), "%s%s", _PATH_HIDE, game);
	if (priority > 0)	/* < 0 requires root */
		(void)setpriority(PRIO_PROCESS, 0, priority);
	execv(pbuf, args);
	err(1, "%s", pbuf);
}

/*
 * read_config --
 *	read through config file, looking for key words.
 */
static void
read_config(void)
{
	FILE *cfp;
	char lbuf[BUFSIZ], f1[40], f2[40], f3[40], f4[40], f5[40];

	if (!(cfp = fopen(_PATH_CONFIG, "r")))
		return;
	while (fgets(lbuf, sizeof(lbuf), cfp))
		switch (*lbuf) {
		case 'b':		/* badtty */
			if (sscanf(lbuf, "%39s%39s", f1, f2) != 2 ||
			    strcasecmp(f1, "badtty"))
				break;
			c_tty(f2);
			break;
		case 'g':		/* game */
			if (sscanf(lbuf, "%39s%39s%39s%39s%39s",
			    f1, f2, f3, f4, f5) != 5 || strcasecmp(f1, "game"))
				break;
			c_game(f2, f3, f4, f5);
			break;
		case 't':		/* time */
			if (sscanf(lbuf, "%39s%39s%39s%39s", f1, f2, f3, f4) != 4 ||
			    strcasecmp(f1, "time"))
				break;
			c_day(f2, f3, f4);
		}
	(void)fclose(cfp);
}

/*
 * c_day --
 *	if day is today, see if okay to play
 */
static void
c_day(const char *s_day, const char *s_start, const char *s_stop)
{
	static const char *const days[] = {
		"sunday", "monday", "tuesday", "wednesday",
		"thursday", "friday", "saturday",
	};
	static struct tm *ct;
	int start, stop;

	if (!ct)
		ct = localtime(&now);
	if (strcasecmp(s_day, days[ct->tm_wday]))
		return;
	if (!isdigit((unsigned char)*s_start) || 
	    !isdigit((unsigned char)*s_stop))
		return;
	start = atoi(s_start);
	stop = atoi(s_stop);
	if (ct->tm_hour >= start && ct->tm_hour < stop) {
		if (start == 0 && stop == 24)
			errx(0, "Sorry, games are not available today.");
		else
			errx(0, "Sorry, games are not available from %s to %s today.",
			     hour(start), hour(stop));
	}
}

/*
 * c_tty --
 *	decide if this tty can be used for games.
 */
static void
c_tty(const char *tty)
{
	static int first = 1;
	static char *p_tty;

	if (first) {
		p_tty = strrchr(gametty, '/');
		first = 0;
	}

	if (!strcmp(gametty, tty) || (p_tty && !strcmp(p_tty, tty)))
		errx(1, "Sorry, you may not play games on %s.", gametty);
}

/*
 * c_game --
 *	see if game can be played now.
 */
static void
c_game(const char *s_game, const char *s_load, const char *s_users, 
       const char *s_priority)
{
	static int found;

	if (found)
		return;
	if (strcmp(game, s_game) && strcasecmp("default", s_game))
		return;
	++found;
	if (isdigit((unsigned char)*s_load) && atoi(s_load) < load())
		errx(0, "Sorry, the load average is too high right now.");
	if (isdigit((unsigned char)*s_users) && atoi(s_users) <= users())
		errx(0, "Sorry, there are too many users logged on right now.");
	if (isdigit((unsigned char)*s_priority))
		priority = atoi(s_priority);
}

/*
 * load --
 *	return 15 minute load average
 */
static double
load(void)
{
	double avenrun[3];

	if (getloadavg(avenrun, sizeof(avenrun)/sizeof(avenrun[0])) < 0)
		err(1, "getloadavg() failed");
	return (avenrun[2]);
}

/*
 * users --
 *	return current number of users
 *	todo: check idle time; if idle more than X minutes, don't
 *	count them.
 */
static int
users(void)
{
	struct utmpentry *ep;
	int nusers;

	nusers = getutentries(NULL, &ep);
	return nusers;
}

static void
nogamefile(void)
{
	int fd, n;
	char buf[BUFSIZ];

	if ((fd = open(_PATH_NOGAMES, O_RDONLY, 0)) >= 0) {
#define	MESG	"Sorry, no games right now.\n\n"
		(void)write(2, MESG, sizeof(MESG) - 1);
		while ((n = read(fd, buf, sizeof(buf))) > 0)
			(void)write(2, buf, n);
		exit(1);
	}
}

/*
 * hour --
 *	print out the hour in human form
 */
static const char *
hour(int h)
{
	static const char *const hours[] = {
	    "midnight", "1am", "2am", "3am", "4am", "5am",
	    "6am", "7am", "8am", "9am", "10am", "11am",
	    "noon", "1pm", "2pm", "3pm", "4pm", "5pm",
	    "6pm", "7pm", "8pm", "9pm", "10pm", "11pm", "midnight" };

	if (h < 0 || h > 24)
		return ("BAD TIME");
	else
		return (hours[h]);
}

#ifdef LOG
/*
 * logfile --
 *	log play of game
 */
static void
logfile(void)
{
	struct passwd *pw;
	FILE *lp;
	uid_t uid;
	int lock_cnt;

	if (lp = fopen(_PATH_LOG, "a")) {
		for (lock_cnt = 0;; ++lock_cnt) {
			if (!flock(fileno(lp), LOCK_EX))
				break;
			if (lock_cnt == 4) {
				warnx("log lock");
				(void)fclose(lp);
				return;
			}
			sleep(1);
		}
		if (pw = getpwuid(uid = getuid()))
			fputs(pw->pw_name, lp);
		else
			fprintf(lp, "%u", uid);
		fprintf(lp, "\t%s\t%s\t%s", game, gametty, ctime(&now));
		(void)flock(fileno(lp), LOCK_UN);
		(void)fclose(lp);
	}
}
#endif /* LOG */
